ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java

```diff
@@ -854,31 +946,37 @@ private StoreFindToken findEntriesFromSegmentStartOffset(Offset initialSegmentSt
       logger.trace(
           "Index : " + dataDir + " findEntriesFromOffset segment start offset " + segmentStartOffset + " with key "
               + key + " total entries received " + messageEntries.size());
-      segmentStartOffset = indexes.higherKey(segmentStartOffset);
-      segmentToProcess = indexes.get(segmentStartOffset);
+      segmentStartOffset = indexSegments.higherKey(segmentStartOffset);
+      segmentToProcess = indexSegments.get(segmentStartOffset);
     }
 
     while (findEntriesCondition.proceed(currentTotalSizeOfEntries.get(), segmentToProcess.getLastModifiedTime())) {
       // Check in the journal to see if we are already at an offset in the journal, if so get entries from it.
       Offset journalFirstOffsetBeforeCheck = journal.getFirstOffset();
       Offset journalLastOffsetBeforeCheck = journal.getLastOffset();
       List<JournalEntry> entries = journal.getEntriesSince(segmentStartOffset, true);
+      Offset endOffsetOfSnapshot = getCurrentEndOffset(indexSegments);
       if (entries != null) {
         logger.trace("Index : " + dataDir + " findEntriesFromOffset journal offset " + segmentStartOffset
             + " total entries received " + entries.size());
         IndexSegment currentSegment = segmentToProcess;
         for (JournalEntry entry : entries) {
           if (entry.getOffset().compareTo(currentSegment.getEndOffset()) > 0) {
-            /* The offset is of the next segment. If the next segment's last modified time makes
-            it ineligible, skip */
-            Offset nextSegmentStartOffset = indexes.higherKey(currentSegment.getStartOffset());
-            currentSegment = indexes.get(nextSegmentStartOffset);
+            Offset nextSegmentStartOffset = indexSegments.higherKey(currentSegment.getStartOffset());
+            // stop if there are no more index segments in this ref.
+            if (nextSegmentStartOffset == null) {
+              break;
+            }
+            currentSegment = indexSegments.get(nextSegmentStartOffset);
+            // stop if ineligible because of last modified time
             if (!findEntriesCondition.proceed(currentTotalSizeOfEntries.get(), currentSegment.getLastModifiedTime())) {
               break;
             }
           }
           newTokenOffsetInJournal = entry.getOffset();
-          IndexValue value = findKey(entry.getKey());
+          IndexValue value =
```

Comment:
We discussed this offline - recording it here.

This situation has been taken care of. The code that precedes this has been added to tackle exactly this situation
```
// stop if there are no more index segments in this ref.
if (nextSegmentStartOffset == null) {
  break;
}
```
Whenever a journal entry that does not exist is encountered, it will have caused a rollover of the index segment. If that index segment does not exist in the world-view that we have, we break out. I have added a comment to call this out explicitly.